File: src/main/java/cookbook/backend/SignUpBackend.java
package cookbook.backend;

public class SignUpBackend {

    // Callback interface to communicate back to the UI
    public interface SignUpListener {
        void onSuccess();
        void onError(String errorMessage);
    }

    /**
     * Initializes any backend connections needed for sign up.
     */
    public static void initConnection() {
        FirebaseManager.connect();
    }

    /**
     * Validates inputs and attempts to sign the user up via Firebase on a background thread.
     */
    public static void attemptSignUp(String email, String password, String confirmPassword, SignUpListener listener) {
        // 1. Basic Validation (No internet needed)
        if (password == null || password.isEmpty() || password.equals("Password")) {
            listener.onError("Please enter a password!");
            return;
        }

        if (!password.equals(confirmPassword)) {
            listener.onError("Passwords do not match!");
            return;
        }

        // 2. Network Request (Run on a background thread)
        new Thread(() -> {
            try {
                String apiKey = System.getenv("FIREBASE_API_KEY"); 
                String resultMessage = FirebaseManager.signUpUser(email, password, apiKey);

                if ("SUCCESS".equals(resultMessage)) {
                    listener.onSuccess();
                } else {
                    listener.onError(resultMessage);
                }
            } catch (Exception e) {
                listener.onError("An unexpected error occurred.");
            }
        }).start();
    }
}

File: app/src/main/java/com/cookbook/backend/SignUpBackend.kt
package com.cookbook.backend

object SignUpBackend {

    sealed class SignUpResult {
        data object Success : SignUpResult()
        data class Error(val message: String) : SignUpResult()
    }

    suspend fun attemptSignUp(email: String, password: String, confirmPassword: String): SignUpResult {
        if (password.isBlank() || confirmPassword.isBlank()) {
            return SignUpResult.Error("Password fields cannot be empty!")
        }
        if (password != confirmPassword) {
            return SignUpResult.Error("Passwords do not match!")
        }
        if (password.length < 6) {
            return SignUpResult.Error("Password must be at least 6 characters!")
        }

        val result = FirebaseManager.signUpUser(email, password)
        return if (result == "SUCCESS") {
            SignUpResult.Success
        } else {
            SignUpResult.Error(result)
        }
    }
}
